src/jit/optimizer.cpp

```diff
@@ -6795,8 +6795,18 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
 
         bool IsTreeVNInvariant(GenTree* tree)
         {
-            return m_compiler->optVNIsLoopInvariant(tree->gtVNPair.GetLiberal(), m_loopNum,
-                                                    &m_hoistContext->m_curLoopVnInvariantCache);
+            ValueNum vn = tree->gtVNPair.GetLiberal();
+
+            if (m_compiler->vnStore->IsVNConstant(vn))
+            {
+                // Don't allow the hoisting of GT_CLS_VAR an GT_LCL_VAR that were assigned a constant value.
+                // But allow constant GT_ARGPLACE nodes which occur when we have constant args.
```

Comment:
I might just request a bit more lengthy comment here. As I understand it, the reason we can't do this thing that seems safe, though perhaps not all that useful, is that we're not checking safety that something that's known to evaluate to a constant is not being modified within the loop. If I have that right, it would be nice to add that here so that people (like me) won't be confused in future when they see this.

Reply:
Is this the right place for this fix? optVNIsLoopInvariant has logic for determining whether the definition is inside the loop. That logic is not executed in this case because of this code:

https://github.com/dotnet/coreclr/blob/ef585a4658a830945f24d635b234f44bd0c81e84/src/jit/optimizer.cpp#L7197-L7201

Perhaps that short-circuit should be removed so that we try to determine if the definition is inside the loop even for const vn's.

Reply:
That function examines the `vn` passed in and sees it it is a  VNF_PhiDef or a VNF_PhiMemoryDef.

So it can only identify the GT_LCL_VARS that have definitions that are tracked by Phi nodes.

When we have a constant `vn` we can't trace back to the definition.

Reply:
optVNIsLoopInvariant also has logic for when the vn is not VNF_PhiDef or a VNF_PhiMemoryDef: it checks function args if vn is a function and it checks vn loop annotations if vn is not a function.

Reply:
That handles things like binary operators such as Add, Sub, etc..
Constant values are terminal node with no children, only a value.

Reply:
> That handles things like binary operators such as Add, Sub, etc..

That part of VN invariance is actually dubious since it's redundant to what the rest of loop hoisting does.

Reply:
One important case is handled here:
We have several "pure" Jit helper functions that take constants as arguments.
It is important that such calls get hoisted into the loop pre-header and then turned into CSE's:

Examples are `CORINFO_HELP_GETSHARED_GCSTATIC_BASE` and  `CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE`

Reply:
Those are already handled in `HoistVisitor::PostOrderVisit` - it already checks for pure helpers with invariant args.

As far as I can tell the only good reason to check VN invariance is to detect memory loads that are invariant. Apart from that the VN invariance check is effectively redundant and confusing:
* `PostOrderVisit` determines that `a` and `b` are invariant and then trivially concludes that `a + b` is also invariant.
* At that point it checks VN invariance - it goes "down" the VN "tree" to check if `a + b` is invariant

Basically as `PostOrderVisit` goes up the expression tree it keeps going down the VN tree again and again attempting to check what has already been checked. If it weren't for `loopVnInvariantCache` this would probably rather slow.

To make things worse, VN invariance could handle the case of a variable defined inside the loop by an invariant expression. Sadly this doesn't happen because `PostOrderVisit` specifically requires that the definition be outside the loop and doesn't check VN invariance if the def is inside the loop.